### ISSUE TITLE
Web process failed to bind to $PORT within 60 seconds of launch

### ISSUE THREAD

**Issue by christiaanwesterbeek** [PostgREST/postgrest-heroku#14]
I created a fresh Heroku app and Amazon RDS instance for Postgres. I can access de Postgres database with some gui (Postico).

For the Heroku app, I followed the procedure described here: https://github.com/begriffs/postgrest-heroku. So: 
```
# first set up a new app
heroku apps:create --buildpack https://github.com/begriffs/postgrest-heroku.git

# now fill in the values specific to your database
heroku config:set POSTGREST_VER=0.4.0.0
heroku config:set DB_URI=postgres://myusername:mypassword@xxx.rds.amazonaws.com/mydbname
heroku config:set DB_SCHEMA=testschema
heroku config:set DB_ANON_ROLE=testrole
heroku config:set DB_POOL=10

git push heroku master
```

Now when I fire up the app, by going to the app url, I get an application error and this is in the log file: 
```
2017-05-24T19:08:49.197400+00:00 heroku[web.1]: Starting process with command `./env-to-config ./postgrest postgrest.conf`
2017-05-24T19:08:50.229037+00:00 app[web.1]: Listening on port 33496
2017-05-24T19:09:22.400036+00:00 heroku[router]: at=error code=H20 desc="App boot timeout" method=GET path="/" host=myappname.herokuapp.com request_id=3382e614-0553-4275-936a-da635589facd fwd="83.162.179.9" dyno= connect= service= status=503 bytes= protocol=https
2017-05-24T19:09:49.667434+00:00 heroku[web.1]: Error R10 (Boot timeout) -> Web process failed to bind to $PORT within 60 seconds of launch
2017-05-24T19:09:49.667434+00:00 heroku[web.1]: Stopping process with SIGKILL
2017-05-24T19:09:49.784127+00:00 heroku[web.1]: Process exited with status 137
2017-05-24T19:09:49.792706+00:00 heroku[web.1]: State changed from starting to crashed
```

With the bind to $PORT being the cause of the crashing app.
Here you can see my config variables ion Heroku:
![image](https://cloud.githubusercontent.com/assets/465989/26421748/69192f80-40c7-11e7-8415-cbd954c6e6cc.png)

I don't know how to debug this as I am new to hosting any kind of other app on Heroku besides Node.js apps. I would sure appreciate some help here.

**Comment by christiaanwesterbeek** [comment]
I get the feeling this is related to this: https://github.com/begriffs/postgrest-heroku/issues/12#issuecomment-283248106

**Comment by begriffs** [comment]
Thanks for the report. The readme here may be out of date with respect to the environment variables you need to set. Try this link and let me know if it works for you: https://heroku.com/deploy?template=https://github.com/begriffs/postgrest

**Comment by christiaanwesterbeek** [comment]
I followed the link, and an error occured during the build:

![image](https://cloud.githubusercontent.com/assets/465989/26442031/20fc056c-4133-11e7-95f7-614f66b83b17.png)
![image](https://cloud.githubusercontent.com/assets/465989/26441932/cce5cef4-4132-11e7-97ab-32b8c5f42cd0.png)
![image](https://cloud.githubusercontent.com/assets/465989/26441968/e8565ab4-4132-11e7-9005-0baf81a5815e.png)
```
-----> DBAPI app detected
-----> Exporting config vars
       DB_ANON_ROLE
       DB_POOL
       DB_SCHEMA
       DB_URI
       POSTGREST_VER
       SECRET_IS_BASE64
-----> Downloading and caching libraries for POSTGREST
-----> Could not find prebuilt postgrest
 !     Push rejected, failed to compile DBAPI app.
 !     Push failed
```

Btw, the configs during the cli-deployment in my first post have the same names as they do during the gui-deployment.

**Comment by begriffs** [comment]
Sorry, I messed up the permissions of the postgrest binary on s3 that gets used for heroku deployment. Give it a try again.

**Comment by christiaanwesterbeek** [comment]
The link you provided is now working. That means the app was deployed successfully. When I open the app though, there's still the same error:

*Error R10 (Boot timeout) -> Web process failed to bind to $PORT within 60 seconds of launch*

Here's the full log that appeared while opening the app for the first time.
```
2017-05-26T05:36:45.190599+00:00 app[api]: Enable Logplex by user chris@...
2017-05-26T05:36:45.190599+00:00 app[api]: Release v2 created by user chris@...
2017-05-26T05:36:45.705725+00:00 app[api]: Release v3 created by user chris@...
2017-05-26T05:36:45.705725+00:00 app[api]: Set BUILDPACK_URL, POSTGREST_VER, DB_URI, DB_SCHEMA, DB_ANON_ROLE, DB_POOL, SECRET_IS_BASE64 config vars by user chris@...
2017-05-26T05:36:46.000000+00:00 app[api]: Build started by user chris@...
2017-05-26T05:36:54.709606+00:00 app[api]: Deploy by user chris@...
2017-05-26T05:36:54.723811+00:00 app[api]: Scaled to web@1:Hobby by user chris@...
2017-05-26T05:36:54.709606+00:00 app[api]: Release v4 created by user chris@...
2017-05-26T05:36:46.000000+00:00 app[api]: Build succeeded
2017-05-26T05:36:56.242479+00:00 heroku[web.1]: Starting process with command `./env-to-config ./postgrest postgrest.conf`
2017-05-26T05:36:58.141188+00:00 app[web.1]: Listening on port 4294
2017-05-26T05:37:56.272860+00:00 heroku[web.1]: Error R10 (Boot timeout) -> Web process failed to bind to $PORT within 60 seconds of launch
2017-05-26T05:37:56.272947+00:00 heroku[web.1]: Stopping process with SIGKILL
2017-05-26T05:37:56.383162+00:00 heroku[web.1]: Process exited with status 137
2017-05-26T05:37:56.400610+00:00 heroku[web.1]: State changed from starting to crashed
```
The app restarts automatically and the failed to bind to $port problem keeps occuring. 

Not sure how to debug these things. But I had a look at https://github.com/begriffs/postgrest-heroku/blob/master/env-to-config and I see this line.

`echo "server-port=${PORT}" >> $CONF_PATH`

Also I found this [on SO](https://stackoverflow.com/questions/15693192/heroku-node-js-error-web-process-failed-to-bind-to-port-within-60-seconds-of): *"Heroku dynamically assigns your app a port, so you can't set the port to a fixed number. Heroku adds the port to the env, so you can pull it from there"*

So, I suspect that somewhere in `env-to-config` the `${PORT}` variable has to be set to `process.env.PORT`.

**Comment by begriffs** [comment]
You're right that env-to-config [adds lines](https://github.com/begriffs/postgrest-heroku/blob/master/env-to-config#L33-L35) at the end of the config to override the server and port because of how heroku assigns random port numbers.

I think the failure to bind from before was that postgrest used to be fragile in how it read fields from the config file. As of postgrest v4.3 (recently released) it is more flexible, so a heroku deploy should now work.

I just tested one a few minutes ago with an RDS backend database and the deploy went smoothly and the app functioned correctly. Could you try with the new version and see if it fixes your problem?

**Comment by christiaanwesterbeek** [comment]
Thanks for your feedback and the improvements. I'm afraid the timeout was probably [due to an Inbound Rule in a Security Group](https://forums.aws.amazon.com/thread.jspa?threadID=158428) attached to my Amazon RDS instance. That rule said to only allow inbound traffic from a specific IP address (being my previous office address). I just found that out today, trying to connect from a different office...

I changed the rule to allow traffic from everywhere. Also I deleted the Heroku app and created a new one using a fresh version of the buildpack (same procedure as I wrote in my previous comment). Not sure which of the 2 changes had the positive effect, but I assume the former.

The api is working perfectly now.
